--- a/ir.c
+++ b/ir.c
@@ -1849,6 +1849,10 @@
 	return FlushInstructionCache(GetCurrentProcess(), ptr, size) == TRUE ? 1 : 0;
 }
 #else
+
+#if defined(__APPLE__) && defined(__aarch64__)
+#include <pthread.h>
+#endif
 
 #if defined(__linux__) && defined(__x86_64__) && defined(PKEY_DISABLE_WRITE)
 # define HAVE_PKEY_MPROTECT 1
@@ -1870,6 +1874,14 @@
 
 void *ir_mem_mmap(size_t size)
 {
+#if defined(__APPLE__) && defined(__aarch64__)
+	void *ret = mmap(NULL, size, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_PRIVATE | MAP_ANONYMOUS | MAP_JIT, -1, 0);
+	if (ret == MAP_FAILED) {
+		return NULL;
+	}
+	pthread_jit_write_protect_np(0);
+	return ret;
+#else
 #ifdef HAVE_PKEY_MPROTECT
 	if (!ir_pkey && pkey_mprotect) {
 		int key = pkey_alloc(0, PKEY_DISABLE_WRITE);
@@ -1901,6 +1913,7 @@
 		ret = NULL;
 	}
 	return ret;
+#endif
 }
 
 int ir_mem_unmap(void *ptr, size_t size)
@@ -1917,6 +1930,10 @@
 
 int ir_mem_protect(void *ptr, size_t size)
 {
+#if defined(__APPLE__) && defined(__aarch64__)
+	pthread_jit_write_protect_np(1);
+	return 1;
+#else
 #ifdef HAVE_PKEY_MPROTECT
 	if (ir_pkey > 0) {
 		if (pkey_set(ir_pkey, PKEY_DISABLE_WRITE)) {
@@ -1935,10 +1952,15 @@
 		return 0;
 	}
 	return 1;
+#endif
 }
 
 int ir_mem_unprotect(void *ptr, size_t size)
 {
+#if defined(__APPLE__) && defined(__aarch64__)
+	pthread_jit_write_protect_np(0);
+	return 1;
+#else
 #ifdef HAVE_PKEY_MPROTECT
 	if (ir_pkey > 0) {
 		if (pkey_set(ir_pkey, PKEY_DISABLE_EXECUTE)) {
@@ -1957,6 +1979,7 @@
 		return 0;
 	}
 	return 1;
+#endif
 }
 
 int ir_mem_flush(void *ptr, size_t size)
